In [2]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score

/Users/alfreed0/opt/anaconda3/envs/geo_solver/lib/python3.9/site-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.3
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


In [3]:
df = pd.read_csv(
    '../Data/cancer.csv'
)

In [4]:
df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

In [21]:
df.columns

Index(['diagnosis', 'radius_mean', 'texture_mean', 'smoothness_mean',
       'compactness_mean', 'concavity_mean', 'concave points_mean',
       'symmetry_mean', 'fractal_dimension_mean', 'radius_se', 'texture_se',
       'smoothness_se', 'compactness_se', 'concavity_se', 'concave points_se',
       'symmetry_se', 'fractal_dimension_se', 'radius_worst', 'texture_worst',
       'smoothness_worst', 'compactness_worst', 'concavity_worst',
       'concave points_worst', 'symmetry_worst', 'fractal_dimension_worst'],
      dtype='object')

In [6]:
df.shape

(569, 33)

In [7]:
df.drop(columns=['Unnamed: 32', 'id'], inplace=True)

In [8]:
df.drop(columns=['area_mean', 'perimeter_mean', 'area_se', 'perimeter_se', 'area_worst', 'perimeter_worst'], inplace=True)

In [9]:
x = df.drop(columns=['diagnosis'])
y = df['diagnosis']

In [10]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.33, random_state=42)

In [11]:
clf = LogisticRegression(random_state=0, max_iter=1000).fit(x_train, y_train)

In [12]:
clf.coef_

array([[-0.57557236, -0.03477011,  0.20923848,  0.5071472 ,  0.84405763,
         0.40333182,  0.31678946,  0.06873389,  1.30264362, -0.31685427,
         0.02894103,  0.08345197,  0.17075484,  0.06660786,  0.05411697,
         0.009665  ,  1.46756277,  0.26999766,  0.34794269,  1.34582384,
         2.07626153,  0.66259878,  0.72396657,  0.16840018]])

In [13]:
clf.predict(x_test)

array(['B', 'M', 'M', 'B', 'B', 'M', 'M', 'M', 'B', 'B', 'B', 'M', 'B',
       'M', 'B', 'M', 'B', 'B', 'B', 'M', 'B', 'B', 'M', 'B', 'B', 'B',
       'B', 'B', 'B', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'B', 'M',
       'B', 'B', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'M',
       'B', 'B', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'M', 'M', 'B', 'B',
       'B', 'M', 'M', 'B', 'B', 'M', 'M', 'B', 'M', 'B', 'B', 'B', 'B',
       'B', 'B', 'M', 'B', 'B', 'M', 'M', 'M', 'M', 'M', 'B', 'B', 'B',
       'B', 'B', 'B', 'B', 'B', 'M', 'M', 'B', 'M', 'M', 'B', 'M', 'M',
       'B', 'B', 'B', 'M', 'B', 'B', 'M', 'B', 'M', 'M', 'B', 'M', 'B',
       'B', 'B', 'M', 'B', 'B', 'B', 'M', 'B', 'M', 'M', 'B', 'B', 'M',
       'M', 'M', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'M', 'B', 'M', 'B',
       'B', 'M', 'B', 'M', 'M', 'M', 'B', 'M', 'B', 'B', 'B', 'B', 'M',
       'M', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'B', 'B', 'M', 'B',
       'M', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'B

In [14]:
clf.score(x_test, y_test)

0.9627659574468085

In [15]:
y_pred = clf.predict(x_test)
y_true = y_test

In [16]:
confusion_matrix(y_true, y_pred)

array([[118,   3],
       [  4,  63]])

In [17]:
accuracy_score(y_true, y_pred)

0.9627659574468085

In [18]:
recall_score(y_true, y_pred, pos_label='M')

0.9402985074626866

In [19]:
y_train.value_counts()

B    236
M    145
Name: diagnosis, dtype: int64

In [20]:
y_test.value_counts()

B    121
M     67
Name: diagnosis, dtype: int64

<font size='10' color='red'>Taller</font>

In [25]:
se_features = df.loc[:, ['diagnosis', 'radius_se', 'texture_se', 'smoothness_se', 'compactness_se', 'concavity_se', 'concave points_se', 'symmetry_se', 'fractal_dimension_se']]

In [26]:
worst_features = df.loc[:, ['diagnosis', 'radius_worst', 'texture_worst', 'smoothness_worst', 'compactness_worst', 'concavity_worst', 'concave points_worst', 'symmetry_worst', 'fractal_dimension_worst']]

In [27]:
x_se = se_features.drop(columns=['diagnosis'])
y_se = se_features['diagnosis']

In [28]:
x_worst = worst_features.drop(columns=['diagnosis'])
y_worst = worst_features['diagnosis']

In [29]:
se_x_train, se_x_test, se_y_train, se_y_test = train_test_split(x_se, y_se, test_size=0.33, random_state=42)

In [30]:
worst_x_train, worst_x_test, worst_y_train, worst_y_test = train_test_split(x_worst, y_worst, test_size=0.33, random_state=42)

In [31]:
se_clf = LogisticRegression(random_state=0, max_iter=1000).fit(se_x_train, se_y_train)

In [32]:
worst_clf = LogisticRegression(random_state=0, max_iter=1000).fit(worst_x_train, worst_y_train)

In [35]:
se_clf.predict(se_x_test)

array(['B', 'M', 'M', 'B', 'B', 'M', 'M', 'M', 'B', 'B', 'B', 'M', 'B',
       'M', 'B', 'B', 'B', 'B', 'B', 'M', 'B', 'B', 'M', 'B', 'B', 'B',
       'B', 'B', 'B', 'M', 'B', 'B', 'M', 'B', 'B', 'B', 'B', 'M', 'M',
       'B', 'B', 'M', 'B', 'B', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'M',
       'B', 'B', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'M', 'M', 'B', 'B',
       'B', 'M', 'M', 'B', 'B', 'M', 'M', 'B', 'B', 'B', 'B', 'B', 'B',
       'B', 'B', 'B', 'B', 'B', 'M', 'B', 'M', 'B', 'M', 'B', 'B', 'B',
       'B', 'B', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'B', 'B', 'M', 'M',
       'B', 'B', 'B', 'M', 'B', 'B', 'M', 'B', 'B', 'M', 'B', 'M', 'B',
       'B', 'M', 'B', 'M', 'B', 'B', 'M', 'B', 'M', 'B', 'B', 'B', 'M',
       'M', 'B', 'B', 'B', 'B', 'M', 'B', 'B', 'M', 'B', 'B', 'M', 'B',
       'B', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'M',
       'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B',
       'B', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'B

In [33]:
se_clf.score(se_x_test, se_y_test)

0.824468085106383

In [36]:
worst_clf.predict(worst_x_test)

array(['B', 'M', 'M', 'B', 'B', 'M', 'M', 'M', 'B', 'B', 'B', 'M', 'B',
       'M', 'B', 'M', 'B', 'B', 'B', 'M', 'B', 'B', 'M', 'B', 'B', 'B',
       'B', 'B', 'B', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'B', 'M',
       'B', 'B', 'M', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'M',
       'B', 'B', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'M', 'M', 'B', 'B',
       'B', 'M', 'M', 'B', 'B', 'M', 'M', 'B', 'M', 'B', 'B', 'B', 'B',
       'B', 'B', 'M', 'B', 'M', 'M', 'M', 'M', 'M', 'M', 'B', 'B', 'B',
       'B', 'B', 'B', 'B', 'B', 'M', 'M', 'B', 'M', 'M', 'B', 'M', 'M',
       'B', 'B', 'B', 'M', 'B', 'B', 'M', 'B', 'M', 'M', 'B', 'M', 'B',
       'B', 'B', 'M', 'B', 'B', 'B', 'M', 'B', 'M', 'M', 'B', 'B', 'M',
       'M', 'M', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'M', 'B', 'M', 'B',
       'B', 'M', 'B', 'M', 'M', 'M', 'B', 'M', 'B', 'B', 'B', 'B', 'M',
       'M', 'B', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'B', 'B', 'M', 'B',
       'M', 'B', 'B', 'B', 'B', 'B', 'B', 'M', 'B', 'B', 'B', 'B

In [34]:
worst_clf.score(worst_x_test, worst_y_test)

0.9680851063829787

In [37]:
se_y_pred = se_clf.predict(se_x_test)
se_y_true = se_y_test

In [38]:
worst_y_pred = worst_clf.predict(worst_x_test)
worst_y_true = worst_y_test

In [39]:
confusion_matrix(se_y_true, se_y_pred)

array([[114,   7],
       [ 26,  41]])

In [40]:
confusion_matrix(worst_y_true, worst_y_pred)

array([[118,   3],
       [  3,  64]])

In [41]:
accuracy_score(se_y_true, se_y_pred)

0.824468085106383

In [42]:
accuracy_score(worst_y_true, worst_y_pred)

0.9680851063829787

In [43]:
recall_score(se_y_true, se_y_pred, pos_label='M')

0.6119402985074627

In [44]:
recall_score(worst_y_true, worst_y_pred, pos_label='M')

0.9552238805970149

<font size='10' color='green'>Analisis</font>

puntuacion modelo se = 0.82
puntuacion modelo worst= 0.96
puntuacion modelo mean= 0.96

Se aprecia que al realizar un modelo de regresion logistica con los datos de la base de datos de cancer de mama por las caracteristicas se, se obtiene una puntuacion de 0.82, lo cual es un valor bajo, por lo que se puede decir que el modelo no es muy bueno, sin embargo, al realizar el modelo con los datos de la base de datos de cancer de mama por caracteristicas worst, se obtiene una puntuacion de 0.96, lo cual es un valor alto, por lo que se puede decir que el modelo es muy bueno. Recordar que la puntuacion es similar al model oque se hizo con el promedio.